File: core/src/de/project/ice/screens/GameScreen.java
package de.project.ice.screens;

import com.badlogic.ashley.core.EntitySystem;
import de.project.ice.IceGame;
import de.project.ice.ecs.IceEngine;
import de.project.ice.utils.DelegatingBlockingInputProcessor;
import org.jetbrains.annotations.NotNull;

import static de.project.ice.config.Config.INVENTORY_KEY;
import static de.project.ice.config.Config.MENU_KEY;

/**
 * Input Handling here...
 */
public class GameScreen extends BaseScreenAdapter
{
    @NotNull
    private final IceEngine engine;
    private final EntitySystem[] SystemsToPause;

    public GameScreen(@NotNull final IceGame game, @NotNull final IceEngine engine)
    {
        super(game);
        this.engine = engine;

        // Load "Scene01" by loading the scene3_load script
        // this.scriptManager.loadScript(scene3_load.class);

        inputProcessor = new DelegatingBlockingInputProcessor(engine.controlSystem)
        {
            @Override
            public boolean mouseMoved(int screenX, int screenY)
            {
                super.mouseMoved(screenX, screenY);
                game.setPrimaryCursor(engine.controlSystem.primaryCursor);
                game.setSecondaryCursor(engine.controlSystem.secondaryCursor);
                return true;
            }

            @Override
            public boolean keyDown(int keycode)
            {
                switch (keycode)
                {
                    case MENU_KEY:
                        GameScreen.this.game.addScreen(new PauseScreen(GameScreen.this.game));
                        return true;

                    case INVENTORY_KEY:
                        GameScreen.this.game.addScreen(new InventoryScreen(GameScreen.this.game));
                        return true;

                    default:
                        return super.keyDown(keycode);
                }
            }
        };
        SystemsToPause = new EntitySystem[]{
                engine.stateSystem,
                engine.cameraSystem,
                engine.animationSystem,
                engine.controlSystem,
                engine.movementSystem,
                engine.scriptingSystem
        };
    }

    @Override
    public void update(float delta)
    {
        engine.update(delta);
    }

    @Override
    public void resume()
    {
    }

    @Override
    public void pause()
    {
    }

    public void pauseGame()
    {
        for (EntitySystem system : SystemsToPause)
        {
            system.setProcessing(false);
        }
    }

    public void resumeGame()
    {
        for (EntitySystem system : SystemsToPause)
        {
            system.setProcessing(true);
        }
    }

    @Override
    public int getPriority()
    {
        return 1000;
    }
}

File: core/src/de/project/ice/screens/GameScreen.kt
package de.project.ice.screens

import com.badlogic.gdx.InputProcessor
import de.project.ice.IceGame
import de.project.ice.config.Config.INVENTORY_KEY
import de.project.ice.config.Config.MENU_KEY
import de.project.ice.ecs.IceEngine
import de.project.ice.utils.DelegatingBlockingInputProcessor

/**
 * Input Handling here...
 */
class GameScreen(game: IceGame, private val engine: IceEngine) : BaseScreenAdapter(game) {
    private val SystemsToPause = arrayOf(
            engine.cameraSystem,
            engine.animationSystem,
            engine.controlSystem,
            engine.movementSystem,
            engine.scriptingSystem,
            engine.breathSystem,
            engine.timeoutSystem,
            engine.andiSystem
    )

    override val inputProcessor: InputProcessor = object : DelegatingBlockingInputProcessor(engine.controlSystem) {
        override fun mouseMoved(screenX: Int, screenY: Int): Boolean {
            super.mouseMoved(screenX, screenY)
            game.primaryCursor = engine.controlSystem.primaryCursor
            game.secondaryCursor = engine.controlSystem.secondaryCursor
            return true
        }

        override fun keyDown(keycode: Int): Boolean {
            if (game.blockInteraction)
                return false

            when (keycode) {
                MENU_KEY, INVENTORY_KEY -> {
                    this@GameScreen.game.addScreen(InventoryScreen(this@GameScreen.game))
                    return true
                }

                else -> return super.keyDown(keycode)
            }
        }
    }

    override fun update(delta: Float) {
        engine.update(delta)
    }

    override fun resume() {
    }

    override fun pause() {
    }

    fun pauseGame() {
        for (system in SystemsToPause) {
            system.setProcessing(false)
        }
    }

    fun resumeGame() {
        for (system in SystemsToPause) {
            system.setProcessing(true)
        }
    }

    override val priority: Int
        get() = 1000
}
